pydra/engine/specs.py

```diff
@@ -66,17 +66,31 @@ class SpecInfo:
 class BaseSpec:
     """The base dataclass specs for all inputs and outputs."""
 
+    def __attrs_post_init__(self):
+        self.files_hash = {}
+        for field in attr_fields(self):
+            if (
+                field.name not in ["_graph_checksums", "bindings", "files_hash"]
+                and field.metadata.get("output_file_template") is None
+            ):
+                self.files_hash[field.name] = {}
+
     def __setattr__(self, name, value):
         """changing settatr, so the converter and validator is run
         if input is set after __init__
         """
-        if inspect.stack()[1][3] == "__init__":  # or name.startswith("_"):
+        if inspect.stack()[1][3] == "__init__" or name in [
+            "inp_hash",
+            "changed",
+            "files_hash",
+        ]:
             super().__setattr__(name, value)
         else:
             tp = attr.fields_dict(self.__class__)[name].type
             # if the type has a converter, e.g., MultiInputObj
             if hasattr(tp, "converter"):
                 value = tp.converter(value)
+            self.files_hash[name] = {}
```

Comment:
i see this is where `precalculated` is reset. just a note that in the future we may want to optimize this further. for example if you have a list of files and i change one or two, it would be nice to not have to recompute the hash of all files. we may want to consider a more global filehash LRU cache that various tasks can dip into.

Reply:
that was my first approach - I had a global dictionary, but than I recalled that you worried that this dictionary would grow too much and I changed to nested dictionary. we can think more later